File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;

        List<Integer> result = new ArrayList<>();

        int startRow = 0;
        int startCol = 0;
        int lastRow = m - 1;
        int lastCol = n - 1;

        int count = m * n;

        while(count > 0){
            // startRow
            for(int i = startCol; i <= lastCol; i++){
                result.add(matrix[startRow][i]);
                count--;
            }
            startRow++;
            if(count == 0) break;

            // last Col
            for(int i = startRow; i <= lastRow; i++){
                result.add(matrix[i][lastCol]);
                count--;
            }
            lastCol--;
            if(count == 0) break;

            // last row
            for(int i = lastCol; i >= startCol; i--){
                result.add(matrix[lastRow][i]);
                count--;
            }
            lastRow--;
            if(count == 0) break;

            // start col
            for(int i = lastRow; i >= startRow; i--){
                result.add(matrix[i][startCol]);
                count--;
            }
            startCol++; 
            if(count == 0) break;
        }

        return result;

    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let m = matrix.length;
    let n = matrix[0].length;

    let sr = 0, ec = n-1, sc = 0, er = m-1;
    let result = []

    let count = m * n;

    while(count > 0){
        // sr
        for(let i = sc; i <= ec; i++){
            result.push(matrix[sc][i]);
            count--;
        }
        sr++;
        if(count == 0) break

        // ec
        for(let i = sr; i <= er; i++){
            result.push(matrix[i][ec]);
            count--
        }
        ec--;
        if(count == 0) break;

        // er
        for(let i = ec; i >= sc; i--){
            result.push(matrix[er][i]);
            count--;
        }
        er--;
        if(count == 0) break;

        //sc
        for(let i = er; i >= sr; i--){
            result.push(matrix[i][sc]);
            count--;
        }
        sc++;
        if(count == 0) break;
    }
    return result;
};